### PR TITLE
Extended `with_columns` to allow **kwargs style named expressions

### PR DESCRIPTION
**Pull request by alexander-beedie**
_Suggested by @kdcd in #3866, but implemented via simple extension of `with_columns`, not as a new function._

-----

Default behavior completely unchanged, but now _also_ allows for the following clean/readable calling convention:

```python
import polars as pl

df = pl.DataFrame({
    "a": [1, 2, 3, 4],
    "b": [0.5, 4, 10, 13],
    "c": [True, True, False, True],
})

# can opt to provide expressions as **kwargs, for 
# additional readability in some common cases.
df.with_columns(
    d = pl.col("a") * pl.col("b"),
    e = pl.col("c").is_not(),
)
# ┌─────┬──────┬───────┬──────┬───────┐
# │ a   ┆ b    ┆ c     ┆ d    ┆ e     │
# │ --- ┆ ---  ┆ ---   ┆ ---  ┆ ---   │
# │ i64 ┆ f64  ┆ bool  ┆ f64  ┆ bool  │
# ╞═════╪══════╪═══════╪══════╪═══════╡
# │ 1   ┆ 0.5  ┆ true  ┆ 0.5  ┆ false │
# ├╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
# │ 2   ┆ 4.0  ┆ true  ┆ 8.0  ┆ false │
# ├╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
# │ 3   ┆ 10.0 ┆ false ┆ 30.0 ┆ true  │
# ├╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┼╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
# │ 4   ┆ 13.0 ┆ true  ┆ 52.0 ┆ false │
# └─────┴──────┴───────┴──────┴───────┘
```
Can freely mix & match with an existing expressions list, or just use one or the other.
All unit tests pass as-is, and some new test coverage was added.

(I think the other idea in that feature request involving `col` also has merit, with a few minor tweaks; however, this PR doesn't touch on that as it is a more fundamental change and likely needs some thought/care ;)

**Update:**

For consistency, the same optional calling pattern is now also allowed with `select` and `agg`.

**Additional update:**

Following discussion, reverted `select` and `agg` updates - they don't read as cleanly as `with_columns`.

**Final update:**

This syntax is considered experimental for now, and requires explicit opt-in via 
```python
pl.Config.set_with_columns_kwargs(True)
```

### CODE REVIEW

**Comment by cbilot** [comment]
One concern that comes to mind: how do we handle kwargs-type syntax if the desired/required column name is not a valid name for a Python keyword?  For example:
```python3
df.with_columns(
    Price Adj = pl.col('stock split') * pl.col('Price'),
    AAPL.Spread = pl.col('AAPL.High') - pl.col('AAPL.Low')
)
```
(The above examples are actual naming conventions from two financial sources I used yesterday.)

Would the user need to mix-and-match kwargs/args syntax, depending on whether the desired/required variable name is a valid name for a python keyword?

**Comment by kdcd** [comment]
I am happy that you liked my suggestion, the only thing I would like to mention is that it would be great to add the same functionality to methods select and agg.

**Comment by cnpryer** [comment]
Any thoughts on maintaining the philosophy that syntactically there should be minimal (ideally one) way to do things?

**Comment by alexander-beedie** [comment]
> Any thoughts on maintaining the philosophy that syntactically there should be minimal (ideally one) way to do things?

There remains one way to do it: `with_columns` ;)

At the high-level yes, you definitely don't want to turn into pandas, with about thirteen ways to do any one thing. However, there is a stark difference between that (where each way probably has a different function, syntax, and wildly different performance characteristics - like how many different ways you can get data out of a frame), and supporting the very common pythonic pattern of `*args` and `**kwargs` in the same function, that execute identically.

**Comment by alexander-beedie** [comment]
> Is there a way to express wild-card expressions using kwargs, such as

Hmm, I don't see an obvious mapping there, unless you have any ideas? The kwargs syntax is a natural extension for the common case, but it's definitely not supposed to be a wholesale replacement.

File: py-polars/polars/cfg.py [hunk 1]
```diff
@@ -105,6 +106,19 @@ def set_global_string_cache(cls) -> type[Config]:
...
         return cls
 
+    @classmethod
+    def set_with_columns_kwargs(
+        cls, enable: Optional[bool] = None
+    ) -> Union[type[Config], bool]:
+        """
+        Enable experimental support for kwargs in `with_columns` method.
+        """
+        if enable is None:
```
Comment (ritchie46) [inline py-polars/polars/cfg.py:116]: I don't think we need an env var here. Maybe we can set a boolean flag on the class? 

This ensure that if multiple users on the same server use polars, this setting doesn't interfere. (I know others do, but as we don't have to cross the FFI boundary here, it is easier to set this in the python running instance).

Comment (alexander-beedie) [reply py-polars/polars/cfg.py]: Done - definitely cleaner.

**Pull request merged**